### ISSUE TITLE
Handle mute/unmute for notifications

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-transition-notify#3]
Planning out the mute/unmute is tricky, so let's document the design issues and implementation challenges in this issue.

**Comment by shankari** [comment]
Current:
- local storage with key = eventName
- retrieves object with key = CONFIG_LIST_KEY
- retrieves list of notifications, each with a unique ID

Option 1: config object
config object similar to data collection, sync, etc
pros:
    if it is key/value, easy to use model similar to other config objects
cons:
    but how do we represent lists.

Option 2: local storage with key = eventName_MUTED
- similar structure, each object has a key MUTED_KEY
- list of notifications, each with structure {"id": ...., "muted": true/false}

Option 3: existing local storage with key = eventName
- have a second object with key = MUTED_LIST_KEY
- list of notifications, similar to option 2

Option 4: current, except that we add a new "muted" flag for each notification

**Comment by shankari** [comment]
Option 3 seems strictly better than option 2 - they both have a separate object
that stores muted/unmuted, but option 3 avoids the kludgy eventName_MUTED key,
and also makes sense that we have the CONFIG_LIST_KEY because we also have a
MUTED_LIST_KEY. Otherwise the CONFIG_LIST_KEY is only required because we can't
store lists in the usercache, only objects

Option 4 seems problematic because we add a listener for the notification at each startup. Storing the muting with the notification would mean that we would override the mute unless we did some more complex updates - e.g. if the notification exists and is muted, then don't override. 

The current implementation will override due to this segment.

```
    } else {
        if ([localNotifyConfig isEqualToDictionary:currList[existingIndex]]) {
            [LocalNotificationManager addNotification:[NSString stringWithFormat:@"configuration unchanged, skipping list modify"]];
            modified = NO;
        } else {
            [LocalNotificationManager addNotification:[NSString stringWithFormat:@"configuration changed, changing object at index %lu", existingIndex]];
            [currList setObject:localNotifyConfig atIndexedSubscript:existingIndex];
        }
    }
```

**Comment by shankari** [comment]
Given all this, Option 3 looks like the best option.
One other issue to think about is that for controlling mute/unmute, we need to display all the configured events/transition pairs so that users can turn notifications on and off.

Since we have a key based on the notification event, we need to have the list of notifications somewhere.

One option is to store it in the javascript interface, or in a separate json file or return it from the interface like we do with the accuracy readings.

Alternatively, we can make the keys something like NOTIFY_eventName instead of eventName and search for all NOTIFY_*. That will make it easier to find all notifications but also increase the chance that people will register for fake notifications.

I'm going to go with the option of storing the list of valid transitions somewhere because the users anyway need to know which are the valid transitions that they can sign up for.

**Comment by shankari** [comment]
One big issue is that it is unclear how best to structure this. For example, do you say that there are configurations for transitions, or do you say that there are two ways of generating notifications - shake and end of trip, and you get to choose which ones you keep?

**Comment by shankari** [comment]
Really tempted to just ignore muting for now and just go with unmuted notifications. And add in muting the next week. Then, I can push to the stores at the end of tomorrow.

**Comment by shankari** [comment]
Note that I need to switch to a different push notification server by end of Jan, which means another push at the end of the week anyway

**Comment by shankari** [comment]
Actually, mid-week!! Yeah that is final then. I am ignoring muting, cleaning up the interface, and pushing to stores tomorrow. Should get started on deploying new server this weekend, not writing new mute code.

**Comment by shankari** [comment]
In response to https://github.com/e-mission/e-mission-transition-notify/issues/3#issuecomment-274241694

> One big issue is that it is unclear how best to structure this. For example, do
> you say that there are configurations for transitions, or do you say that there
> are two ways of generating notifications - shake and end of trip, and you get
> to choose which ones you keep?

It is possibly best to think of this as two separate pieces of functionality.
Right now, we envisage that both pieces of functionality will do the same
thing, so people need to choose one, but that might not always be the case.
For example, maybe people want to do a survey at the end of the trip and use
the shake to report when they are sad, for example.

So there should really be a list of prompts for each gesture. The list is
custom to the app (and can be set via javascript). And the user can
enable/disable anything in the list. Makes sense!

**Comment by shankari** [comment]
Fixed by #9
